.github/actions/cross-tests/action.yml

```diff
@@ -0,0 +1,41 @@
+name: "cross-tests"
+
+inputs:
+  rust:
+    required: true
+  package:
+    required: true
+  target:
+    required: true
+  features:
+    required: true
+
+runs:
+  using: "composite"
+  steps:
+      - uses: actions/checkout@v2
+      - uses: actions-rs/toolchain@v1
+        with:
+          profile: minimal
+          toolchain: ${{ inputs.rust }}
+          target: ${{ inputs.target }}
+          override: true
+      - name: Install precompiled cross
+        run: |
+          export URL=$(curl -s https://api.github.com/repos/cross-rs/cross/releases/latest | \
+            jq -r '.assets[] | select(.name | contains("x86_64-unknown-linux-gnu.tar.gz")) | .browser_download_url')
+          wget -O /tmp/binaries.tar.gz $URL
+          tar -C /tmp -xzf /tmp/binaries.tar.gz
+          mv /tmp/cross ~/.cargo/bin
+        shell: bash
```

Comment:
Curious how much time this actually saves versus `cargo install cross`, which we use elsewhere:

https://github.com/RustCrypto/elliptic-curves/blob/89d6ab4/.github/workflows/k256.yml#L136

Looking at the CI history, it takes about 1m11s

Reply:
Yep, I also thought about using `cargo install cross`. The benefit is not too big for the more complex CI code.

`cross` within my fork takes 30 - 40 seconds for a single run.
https://github.com/aewag/sponges/actions/runs/1745945800

I can push a variant using `cargo install`, if it performs much worse.

Reply:
Well, there seems to be a different issue than performance:

The current `MSRV` of the `keccak` crate is 1.41. But this is not sufficient to build cross [1]. So best is probably to work with the precompiled `cross`. I removed the WIP commit for experimenting with `cargo install`.

[1] https://github.com/aewag/sponges/runs/4940099815?check_suite_focus=true#step:3:170

Reply:
It's ok to bump MSRV, especially if we're making breaking changes anyway

Reply:
Isn't than the MSRV of cross the bottomline MSRV of keccak?

I just pushed a WIP with MSRV 1.46. This succeeds building `cross`. Interestingly prior versions fail [1], although cross MSRV is 1.42. Not really sure why this is the case.

Performance-wise it takes 2m for a single run compared to 30-40s with a precompiled binary.

[1] https://github.com/aewag/sponges/runs/4942085469?check_suite_focus=true

Reply:
Ok, well that seems significant enough to maybe look at creating some centralized install recipe for

Reply:
I think we should replace `cargo install cross` with downloading of pre-compiled binaries, not the other way around. I don't think it makes much sense to re-compile cross from scratch each time, even if time-wise the difference is not too big.

Reply:
The open question here was how to have a single recipe for installing `cross` shared across all of our repos, so we're not copying and pasting boilerplate

Reply:
Maybe it's worth doing on the `cross` side (i.e. `cross` installation action)? I think we are far from the only one who uses it in their CI.

Reply:
I just tried it via `actions-rs/cargo` [1]. It does build `cross` as well [2]. I don't know of any other solution than either building or fetching a pre-compiled binary.

I would suggest to move the GH template action for `cross` into a separate repository within the RustCrypto organization and only fetch it within each pipeline. This could also be extended with further common templates, if required in the future.

[1] https://github.com/actions-rs/cargo
[2] https://github.com/aewag/sponges/runs/4953414203?check_suite_focus=true#step:4:80

Reply:
Hm, we probably should consider relying on `actions-rs/cargo` as described [here](https://github.com/actions-rs/cargo#cross-compilation). After a cursory look, I also failed to determine whether it build `cross` from scratch or uses pre-built binaries.

Reply:
I originally used `actions-rs/cargo` to drive `cross` but as @aewag noted, it does little more than `cargo install cross` so it felt like a needless dependency. Also it seems a bit unmaintained as of late, and if anything I think we should migrate away from it.

I think there are also several other places we could use some shared CI configuration, and that we should perhaps consider making our own repo for shared GH Actions configs.

Reply:
> I think there are also several other places we could use some shared CI configuration, and that we should perhaps consider making our own repo for shared GH Actions configs.

Just to see, how this could look like I played with the `sha2` [CI configuration](https://github.com/aewag/hashes/pull/2/files).
[test for different OSs](https://github.com/aewag/hashes/blob/147873dc885eba159f22c0cb1bdd19412de24214/.github/workflows/test.yml) and [cross](https://github.com/aewag/hashes/blob/147873dc885eba159f22c0cb1bdd19412de24214/.github/workflows/cross.yml) is split out as Reusable Workflow (and could be stored into a separate repo for shared usage).
To define the test vectors a [stringified JSON](https://github.com/aewag/hashes/blob/147873dc885eba159f22c0cb1bdd19412de24214/.github/workflows/sha2.yml#L43) is needed to create a GHA matrix.

Reply:
How difficult it would be to write an `actions-rs/cargo` analog which would install `cross` using pre-compiled binaries without running tests themselves? For example, in the `aes` case `cross` tests are more [involved](https://github.com/RustCrypto/block-ciphers/blob/cipher_v0.4/.github/workflows/aes.yml#L199-L216) and do not quite fit the action you have written.

Reply:
> How difficult it would be to write an `actions-rs/cargo` analog which would install `cross` using pre-compiled binaries without running tests themselves? 

I have no experience with that. But if the goal is to have an action, which only installs `cross` using pre-compiled binaries [this](https://github.com/aewag/hashes/blob/51aa3d50fee6576703c880fb9ce6f0b4215113d0/.github/actions/cross-install/action.yml) should work. I just adapted the [cross.yml](https://github.com/aewag/hashes/blob/51aa3d50fee6576703c880fb9ce6f0b4215113d0/.github/workflows/cross.yml#L30) to call it.

> For example, in the `aes` case `cross` tests are more [involved](https://github.com/RustCrypto/block-ciphers/blob/cipher_v0.4/.github/workflows/aes.yml#L199-L216) and do not quite fit the action you have written.

Yes, true. It may make sense to split this issue. So one thing is to get `cross` without having to compile it and the other if the tests should be based on shared design.

IMO the first should be fine by using CompositeActions as [above](https://github.com/aewag/hashes/blob/51aa3d50fee6576703c880fb9ce6f0b4215113d0/.github/actions/cross-install/action.yml). For the second, having a shared design, utilizing ReusableWorkflows could work. The `cross` workflow could be for example [extended](https://github.com/aewag/hashes/blob/51aa3d50fee6576703c880fb9ce6f0b4215113d0/.github/workflows/cross.yml#L31-L32) to have the option for `env.RUSTFLAGS` as required for the `aes` case. (The question is how much more is needed.)

Reply:
@tarcieri
We could, but I wonder whether it's worth the trouble for just saving 5 lines of boilerplate in rarely touched CI configs. Is there anything else which could go into this repo? Testing on different OSes is less important for us since most crates are `std`-agnostic.

Reply:
Perhaps @aewag can check me on this but I think the `set-msrv` pattern introduced in this PR (and also [this one, which now has differing cross-install boilerplate](https://github.com/RustCrypto/hashes/pull/339/files)) could potentially be extracted as a reusable workflow?

Even with 5 lines of boilerplate, it's definitely something it'd be nice to update in a single place as opposed to having to replace some copypasta across dozens of repos.

Reply:
> Perhaps @aewag can check me on this but I think the `set-msrv` pattern introduced in this PR (and also [this one, which now has differing cross-install boilerplate](https://github.com/RustCrypto/hashes/pull/339/files)) could potentially be extracted as a reusable workflow?

Yes, `set-msrv` is a candidate to be moved into a reusable workflow. For example, this would easily allow an extension  to get the MSRV from the [`Cargo.toml`](https://doc.rust-lang.org/cargo/reference/manifest.html#the-rust-version-field) instead of hardcoding it in the CI config. I think for such extensions it is nice to have common patterns at a single place.

 
> Additionally it seems like actions-rs is unmaintained and a rather heavyweight solution for how we actually use it. Perhaps we could fork it and slim it down to our specific needs. That would eliminate dependence on third-party actions we don't control.

I suggest to install `cross` using a [composite action](https://github.com/aewag/hashes/blob/51aa3d50fee6576703c880fb9ce6f0b4215113d0/.github/actions/cross-install/action.yml). This seems to me the most lightweight approach, that is currently available. If `actions-rs` gets extended or `cross` supports better options to install, we could simply switch to that by modifying this action only.

Reply:
@newpavlov WDYT? Should we give it a try?

Reply:
@tarcieri 
Yeah, go ahead. We always can revert everything back if it will be too much trouble in the end.

Reply:
I made a repo here: https://github.com/RustCrypto/actions

@aewag feel free to upstream the work there! It seems like we'll be able to unify the configs in `sponges` and `hashes`

Reply:
> I made a repo here: https://github.com/RustCrypto/actions
> 
> @aewag feel free to upstream the work there! It seems like we'll be able to unify the configs in `sponges` and `hashes`

Perfect, we'll do that. :+1: